### DISCUSSION TITLE
Included this project in a website, would be awesome to eventually make it built in

### DISCUSSION THREAD

**Discussion by lunarnautics** [OpenBB-finance/OpenBB#565] [Ideas]
Hey everyone! I recently created an open source website as a source of education and resources for people who want to learn more about finance (such as all of us). It is still under development and anyone is able to contribute. The site can be found here:

https://thehitchhikersguidetofinance.github.io/

I have added this terminal as a link, and I think it would be very cool if it can be integrated in the site in the future. I noticed there was some work being done on a UI and I am curious how easy it would be to include this with html somehow?

**Comment by lunarnautics** [comment]
Of course API keys would need to be worked out. Any ideas on this too?

**Reply by fry-day** [reply]
It is not currently possible to include GST in a static website (or more specifically, a statically-hosted website), such as the one you have built using Hugo.

GST requires a dynamic runtime environment, one that can run Python code. An example of a host that provides this type of dynamic server environment would be Linode or DigitalOcean (there are many more).

The next step to making GST available on a web site would be to create an HTTP interface for it. Right now GST only has a command-line interface (with a desktop GUI in the works). Neither of these would be directly accessible remotely from a browser, even if you successfully set up GST on a server. Creating an HTTP interface would involve a fair amount of work, including adding a concept of sessions (and probably users & authentication as well).

More about sessions: Right now GST assumes only 1 person is using it, or in other words, there is only 1 session. If you exposed an HTTP interface naively, it would be as if anyone could "hijack" that session and start making changes, and it would effect anyone else who was trying to use it. By adding a concept of sessions, you could allow different people to be doing different things (looking at different stocks, running different commands, each doing their own thing).

In addition to turning GST in to a web service (with an HTTP API and sessions at least), you would then need to build a web "front end". This can be handled in two ways:

1. Return HTML directly from the dynamic back-end server, to be rendered by the browser. A framework like Django could be helpful for building this (providing many of the building-blocks typical web apps need). This is a bit like WordPress, in case you are familiar with that, except more general purpose. This could be embedded in your static website via an iframe.
2. Return raw data (JSON) from the dynamic back-end server, and create a dynamic front-end, using Javascript that runs right in the browser. This Javascript code would interact with the dynamic back-end server and be responsible for generating/updating the HTML page elements that the browser renders. View libraries like Angular or React are commonly used to build dynamic front-end apps. Because Javascript can run in the browser, it would be possible to include the Javascript app directly in your statically-hosted website (and most Javascript web-apps _are_ served/hosted statically, for performance reasons, even if they interact with a dynamic API/back-end).

Regarding API keys: because the dynamic application code would be running in a protected server environment (only accessible through the HTTP interface), the API keys used to access external sources would be private from outside eyes.

This would be a fantastic long-term goal (making GST available as a service), however you would pretty quickly run in to the issue of cost. Dynamic servers require _much_ more computing resources than static websites, which can get pricy (especially when you add in database servers, required by most web apps to "persist" data between requests). At that point, GST would most likely need to charge for this service (unless sponsored somehow). Perhaps a "freemium" service could be viable (where free users get access to basic features, and paid users access premium features while covering the overhead costs for everyone).

**Reply by lunarnautics** [reply]
Thank you for a great explanation. That does sound more like a long term goal. I am thinking for now the best way forward will be to integrate data generated from it. Things like this:

https://htmlpreview.github.io/?https://github.com/GamestonkTerminal/GamestonkTerminal/blob/main/notebooks/examples/gme_20210726_030502_due_diligence.html

**Reply by lunarnautics** [reply]
> 

I'm curious, do you think something like this could be used?

https://leaningtech.com/webvm-server-less-x86-virtual-machines-in-the-browser/

**Comment by CoconutMacaroon** [comment]
@lunarnautics I don't know if you're still interested, but [I created a version of Gamestonk Terminal that runs in the browser](https://github.com/GamestonkTerminal/GamestonkTerminal/pull/1251). It isn't static, but if you're still interested, check it out.

**Reply by lunarnautics** [reply]
Thanks for showing me! I am still interested. I quickly searched how to run python code through html and that might be an option. I'd like to avoid servers if possible

**Reply by CoconutMacaroon** [reply]
I'm not sure how you would run the Python code without a server. Not saying you can't, just I don't know. Of course, by doing that, you would be trusting the client device with all the API keys (if you use any), so you'd want to factor that in as well.

**Reply by JakeRuth** [reply]
We are building https://stockunlock.com/ as fellow stonk/coding nerd... I think it's what you're looking for check out the free form tool (you need an account but you can use a fake email and we don't verify anything/totally free to try and use with no bs so no worries there mate)
